File: leetcode/0091-decode-ways.cc
// https://leetcode.com/problems/decode-ways/
// Medium

#include<unordered_map>
#include<string>
#include<iostream>
using namespace std;

class Solution {
    int numDecodingsMemo(string s, unordered_map<string, int>& m){
        if (m.find(s) != m.end()) {
            return m[s];
        }
        int len = s.length();
        if (s == "0" || s == "00") {
            return 0;
        }
        if (len == 1) {
            return 1;
        }

        if (s[0] == '0') return 0;
        int sum = numDecodingsMemo(s.substr(1, len), m);
        string temp = s.substr(0, 2);
        if (temp <= "26") {
            if (len - 2 > 0) {
                int x = numDecodingsMemo(s.substr(2, len), m);
                if (x != 0) {
                    sum += x;
                }
            } else {
                sum ++;
            }
        }
        m[s] = sum;
        return sum;
    }
public:
    int numDecodings(string s) {
        unordered_map<string, int> m;
        return numDecodingsMemo(s, m);
    }
};

int main() {
    string s = "26";
    string s1 = "27";
    cout << (s1 > "19") << endl;
}

File: leetcode/python/0091-decode-ways.py
# https://leetcode.com/problems/decode-ways/
# Medium

# Create a set that contains all the mappings from string 1..26
# Go backwards (or forwards) in the string, at each point checking if the
#  current character or the two consecutive characters belong in the set
# If they do, add the dp value of the corresponding index
class Solution:
    def numDecodings(self, s: str) -> int:
        d = set(str(i) for i in range(1, 27))
        dp = [0]*(len(s)+1)
        dp[-1] = 1
        dp[-2] = 1 if s[-1] != '0' else 0
        for i in range(len(s)-2, -1, -1):
            if s[i:i+1] in d:
                dp[i] += dp[i+1]
            if s[i:i+2] in d:
                dp[i] += dp[i+2]
        return dp[0]
        